### PR TITLE
Implement a follow_next_links method

### PR DESCRIPTION
**Pull request by DevLazio**
Related : https://github.com/sreeise/graph-rs/discussions/360
https://github.com/sreeise/graph-rs/issues/362

### CODE REVIEW

**Comment by DevLazio** [comment]
Still need to update the documentation

**Comment by sreeise** [comment]
I will try to look at this here in the next couple of days.

File: graph-http/src/traits/link.rs
Comment (sreeise) [inline graph-http/src/traits/link.rs:43]: You should either map this to an Option or return it as a Result. We don't know if this value will be there and we don't want to unwrap without a very good reason. I usually only ever do so when dealing with threads because that is common practice in Rust.

File: tests/async_concurrency.rs
Comment (sreeise) [inline tests/async_concurrency.rs:9]: We can probably name this better. Something more related to what it does. `odata_next_link` possibly.

File: graph-http/src/intoresponse.rs
Comment (sreeise) [inline graph-http/src/intoresponse.rs:332]: Similar to how you added the method for setting the next link in the RequestClient trait we should add a method to get that value and call it her instead. I try to keep all of the places where we are calling `lock()` in only a few files so that we know where to find them and can debug them easier. `IntoResponse` shouldn't have any place that is calling the lock() method. I don't remember if there are places in there where its being done but if there is that is something that a ticket should get created for to fix it.

Adding a method with something like:
```rust
if self.client.should_follow_next_links()
```
or
```rust
if self.client.get_follow_next_links()
```
or
```rust
if self.client.following_next_links()
```

something like that.

File: graph-http/src/intoresponse.rs
Comment (sreeise) [inline graph-http/src/intoresponse.rs:178]: We are gonna need a better solution then requiring every caller to implement ODataNextLink<V> just to call the json method. Yes the trait is implemented for serde_json::Value but not for user defined types. That would reduce functionality instead of increasing it and should probably be a different method.

File: graph-http/src/intoresponse.rs
Comment (sreeise) [inline graph-http/src/intoresponse.rs:207]: One of my concerns about this is that there may be hundreds to thousands of next link calls which will cause a very long delay, blocking the thread for an unknown time, and making it difficult to use effectively in production environments in those cases. We should add a doc comment over this method to tell the caller that this method could result in a long running operation when there are many next link calls.

**Review by sreeise — CHANGES_REQUESTED** [review]
Added my initial thoughts. Will take a look again here in the next couple days.

File: graph-http/src/traits/link.rs
Comment (DevLazio) [reply graph-http/src/traits/link.rs:43]: Yeah, good idea. 
It will solve the unwrap and will be much easier to implement.

File: graph-http/src/intoresponse.rs
Comment (DevLazio) [reply graph-http/src/intoresponse.rs:178]: I ran into this problem when updating doc and example. 
That's why I added an impl in the proc macro for the singular type but I agree that we could do something better.

I wanted to implement it for all T but it's not possible unfortunately.

Comment (sreeise) [reply graph-http/src/intoresponse.rs:179]: Im struggling to figure out what the V in ODataNextLink<V> does in this situation? I dont see `Vec<V>` returned in the response so I am wondering what its use case is. Also, shouldnt this return the `json_value Vec<V>` since that includes all of the requests.?

File: graph-http/src/intoresponse.rs
Comment (sreeise) [inline graph-http/src/intoresponse.rs:177]: Good job with the README comments. I was more referring to a method comment though.

File: graph-macro/src/lib.rs
Comment (sreeise) [inline graph-macro/src/lib.rs:15]: I wonder if this might be better as a derive macro.

**Review by sreeise — CHANGES_REQUESTED** [review]

File: graph-http/src/intoresponse.rs
Comment (sreeise) [inline graph-http/src/intoresponse.rs:319]: It might be better to just use a struct that is a placeholder for T/phantom data.

This way you could have specific responses only for T if T is ODataNextLink

```rust
impl<'a, NextLink<T>> IntoResponseAsync<'a, NextLink<T>> {
    pub async fn json<T>>(self) -> **GraphResult<GraphResponse<U>>**
```

Something like that.

**Review by sreeise — CHANGES_REQUESTED** [review]

File: graph-macro/src/lib.rs
Comment (DevLazio) [reply graph-macro/src/lib.rs:15]: Semanticaly, derive macro are expected to only implement a trait for the type, even if it's not a hard rule.


At first I went for the proc_macro_attribute instead of the derive macro because derive macro  can't modify the item but in the end I didn't need that capability. I can try to change that to a derive macro if you want.

File: graph-http/src/intoresponse.rs
Comment (DevLazio) [reply graph-http/src/intoresponse.rs:179]: V is the type defined by the user, that is included in U as a Vec<V>. Unfortunately I didn't find a way to express that relationship for the compiler without a templated trait.


```
if let Some(json_value) = json.value() {

                json_value.append(&mut values);

            }
```
This part takes care of putting back all the requests data into the first request.

File: graph-http/src/intoresponse.rs
Comment (DevLazio) [reply graph-http/src/intoresponse.rs:319]: I think I understand what you mean. Not sure how to do that but I can give it a go.


Would this not result in a "multiple definitions" for json() ?

Comment (DevLazio) [reply graph-http/src/intoresponse.rs:319]: I tried to implement that as 
``` rust 
#[derive(Deserialize)]
pub struct NextLink<'a, V>
{
    value: Option<&'a mut Vec<V>>,
    next_link: Option<String>,
}

impl<V> NextLink<'_, V> {
    fn next_link(&self) -> Option<String> {
        self["@odata.nextLink"].as_str().map(|s| s.to_string())
    }

    fn value(&mut self) -> Option<&mut Vec<V>> {
        self["@odata.value"].as_array_mut()
    }
}

impl<'a, T> IntoResponseAsync<'a, NextLink<'_, T>> {
pub async fn json<U>(self) -> GraphResult<GraphResponse<U>> {

```
and the :crab: is not happy :
```
error[E0592]: duplicate definitions with name `json`
   --> graph-http\src\intoresponse.rs:334:1
    |
334 |   pub async fn json<U>(self) -> GraphResult<GraphResponse<U>>
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `json`
...
376 | /     pub async fn json<U>(self) -> GraphResult<GraphResponse<U>>
377 | |         where
378 | |                 for<'de> U: serde::Deserialize<'de>,
    | |____________________________________________________- other definition for `json`
```

Am I missing something ?

Comment (sreeise) [reply graph-http/src/intoresponse.rs:319]: I was meaning more like how NoContent is used. But I gave the wrong example before

https://github.com/sreeise/graph-rs/blob/248f9298551e721042b7ee9a27e99ede3e5f0cce/graph-http/src/types/content.rs#L2

https://github.com/sreeise/graph-rs/blob/248f9298551e721042b7ee9a27e99ede3e5f0cce/graph-http/src/intoresponse.rs#L244

Its a struct that is basically a placeholder so that only request methods that return NoContent can use that specific implementation.

I was thinking with your method with_next_links (or whatever its called) you would return IntoResponse but with a different type. The method would still be part of the generic impl of IntoResponse but then it would return a more concrete type of IntoResponse:

```rust
#[derive(Clone, Debug, Default)]
pub struct NextLink;


impl<'a, T> IntoResponseBlocking<'a, T> {
    pub fn with_next_links(self) -> IntoResponseBlocking<'a, NextLink> {
        self
    }
}

impl <'a> IntoResponseBlocking<'a, NextLink> {
    pub fn json<U>(self) -> GraphResult<GraphResponse<U>>
        where
                for<'de> U: serde::Deserialize<'de>,
    {
         // ....
    }
}
```

This should prevent any issues with duplicate definitions since the type of T is concrete and has its own implementation where you can have a separate json method.

Comment (sreeise) [reply graph-http/src/intoresponse.rs:319]: Based on the above you could possibly try

```rust
impl<'a, T> IntoResponseBlocking<'a, T> {
    pub fn with_next_links(self) -> IntoResponseBlocking<'a, NextLink<'a, T>> {
        IntoResponse::new(self.client)
    }
}

impl<'a, T> IntoResponseBlocking<'a, NextLink<'a, T>> {
    pub fn send(self) -> GraphResult<GraphResponse<T>>
        where
                for<'de> T: serde::Deserialize<'de> + ODataNextLink,
    {
     /// ...
    }
}
```

And when calling its something like:

```rust
client.v1()
  .users()
  .list_user()
  .with_next_links::<serde_json::Value>()
  .send()
  .unwrap();

```

Comment (DevLazio) [reply graph-http/src/intoresponse.rs:319]: It does indeed work when doing it like that.
Any reason that json() is so generic and not restricted to `T: serde::Deserialize` ?
Neither my work project nor tests fail when I move the `json()` to the 
```rust
impl<'a, T> IntoResponseBlocking<'a, T>
where
    for<'de> T: serde::Deserialize<'de>,
{
}
``` 
block.

We have 3 possibilities :
- name it `send`, but the naming does not follow the rest of the library
- name it `json`, but implies to restrict the generic method
- name it something like `json_with_next_links` and remove the `with_next_links_calls` method

Comment (sreeise) [reply graph-http/src/intoresponse.rs:319]: Nope, no specific reason for that outer T being generic except for the other two methods there, text and bytes, which dont need to implement Deserialize.

So you could move the json method to another implementation block with the T restricted to Deserialize and just use that. The text and bytes methods can just stay in the same place they are now with that generic T.

File: graph-macro/src/lib.rs
Comment (sreeise) [reply graph-macro/src/lib.rs:15]: I guess I was thinking more derive macro because you could set attributes for a next link field given any struct. 

```rust
#[derive(ODataLink)]
pub struct Users {
  #[ODataLink(odata_next_link)]
  pub next_link: Option<String>
  #[ODataLink(odata_value)]
  pub value: Option<string>
}
```

And then use those fields.

**Review by sreeise — CHANGES_REQUESTED** [review]

File: graph-macro/src/lib.rs
Comment (DevLazio) [reply graph-macro/src/lib.rs:15]: We could provide a derive macro for users who want to declare their struct `Users` instead using one provided by the odata_next_link proc macro (on the `User` struct) , but it won't replace it.

I started working on it a bit and it's not trivial at all : we need to find the correct attribute for value (either by name or with the custom macro attribute) and find the type inside the Vec. But we need to handle every cases : `Vec<User>`,  `std::vec::Vec<User>`, `Vec<User<T>>` , etc.

File: graph-http/src/intoresponse.rs
Comment (DevLazio) [reply graph-http/src/intoresponse.rs:408]: I have no idea what clippy is talking about. I can compile it fine and my local clippy does not say that.

Comment (sreeise) [reply graph-http/src/intoresponse.rs:408]: Ok ill pull down your branch locally and see if I can find out whats going on. I will probably just dismiss the alerts if I cant find anything.

**Comment by sreeise** [comment]
So the issue was basically in the master merge itself for the PR. Somehow the old ODataNextLink was still in the branch after merging which was causing clippy to show those linting errors.

**Comment by sreeise** [comment]
I had to just remove automatic failures for clippy and just do default warnings. Its somehow always manages to find something wrong that it couldnt find locally 🤦 its not even worth it 😆

**Comment by sreeise** [comment]
Ok so the rust build failure is just because of permissions so its all good. Tests are passing for me locally. I am looking at your changes now and going over them. I will try to get back to you soon.

**Comment by sreeise** [comment]
@DevLazio 

I can confirm that this does work as intended. Good job getting all this together. This is really good work. 

I like that we can just do paging in one call:

```rust
let result: GraphResult<GraphResponse<serde_json::Value>> = client
    .v1()
    .users()
    .list_user()
    .with_next_links_calls()
    .json()
    .await;
```

Thats a cool way of doing it. This is gonna work really well for many of us who just want to get all of the paging results without having to use a more complicated solution such as message passing. 

On using the macro and why its needed I am not sure that its something that makes sense from a client library use case perspective. Its not that I dont think it can be useful to some people but that the way in which it does it makes it more complicated than it needs to be. 

Lets look at an example:

```rust
#[derive(Debug, Serialize, Deserialize)]
#[odata_next_link]
pub struct User {
    pub(crate) id: Option<String>,
    #[serde(rename = "userPrincipalName")]
    user_principal_name: Option<String>,
}

let result: GraphResult<GraphResponse<Users>> = client
    .v1()
    .users()
    .list_user()
    .select(&["id", "userPrincipalName"])
    .with_next_links_calls()
    .json()
    .await;
``` 

This example is doing couple things as you already know.

* Creates a new struct called Users
* Defines a field on the struct for the next link
* Defines a field that will be a list of the resource, in this case User, that we want to get.

Heres where problems can start to arise:

* We are starting to wade into the waters of creating and implementing types for the user instead of providing a mechanism by which they can use their own types.
* Conflicts where the user of the library has their own Users struct, or whatever name is provided, already and has a different use case for it.

I think this is also a little bit of over engineering when we can just use return a `Vec<T>` whereby the T is required to implement `ODataNextLink` and the `ODataNextLink` just has the `next_link()` method. 

I always prefer simple solutions unless the value that a solution provides out weighs the cost of a complicated one. But in this case I am not sure it justifies doing so.

I am open to listening why you think this would be a better solution.

**Comment by DevLazio** [comment]
> 

That's a great summary.

If I remember right, we went this way because 1) we wanted to be able to return the others fields associated with the response and 2) to not introduce a breaking change in the API.
The second part is no longer relevant thanks to the NextLink specialisation.

As a purely personal point of vue, I've never needed these other fields so I'm totaly fine with returning a `Vec<T` and not bothering with the proc macro, but my use cases are limited.

**Comment by sreeise** [comment]
> > 
> 
> That's a great summary.
> 
> If I remember right, we went this way because 1) we wanted to be able to return the others fields associated with the response and 2) to not introduce a breaking change in the API. The second part is no longer relevant thanks to the NextLink specialisation.
> 
> As a purely personal point of vue, I've never needed these other fields so I'm totaly fine with returning a `Vec<T` and not bothering with the proc macro, but my use cases are limited.

Yeah I see what your saying. I think in this case the caller could really just include or not include whatever fields they want in their own struct. Or even just use the select method.

Yeah, lets get remove the proc macro and just return a Vec<T>. I think that will also be simpler from a user perspective as well. 

I also think we should probably rename the `with_next_links_method` to `paging` because that is what its called by the graph api and it makes more sense I believe. So it would look like:

```rust
let result: GraphResult<GraphResponse<serde_json::Value>> = client
    .v1()
    .users()
    .list_user()
    .paging()
    .json()
    .await;
```

**Comment by DevLazio** [comment]
Alrights, I'll do that.

I'm not a native english speaker, but I would assume that a `paging` method would return some kind of stream or collection of collection. Something like `flatten_paging`  or `ignore_paging` ? I'm really bad at naming though.

**Comment by sreeise** [comment]
> Alrights, I'll do that.
> 
> I'm not a native english speaker, but I would assume that a `paging` method would return some kind of stream or collection of collection. Something like `flatten_paging` or `ignore_paging` ? I'm really bad at naming though.

Yeah your right. Its basically returning a subset of all of the data and paginating just allows you to continue getting the data. Which is what the next links are doing except we are just returning it as a whole. You could honestly just call it `flatten()` if that sounds good to you. 

```rust
let _ = client
    .v1()
    .users()
    .list_users()
    .paging()
    .flatten();
```

Eventually we can add multiple ways of doing this such as getting a stream.

**Comment by DevLazio** [comment]
> This would be a `flatten<T>()` technically instead of the` json<T>()` so I dont know if you had your mind set on wanting it to be called `json`.
> 

Well, I liked the consistency and hiding of the implementation details between the "standard" `json` method and the "paged" `json` method. If that's okay with you.

My branch should be up to date with the lastest requested changes :
- removing the macro
- returning the GraphResponse<Vec<T>>
- renaming `with_next_links_calls` to `paging`
- removing now unused methods, attributes and trait

**Review by sreeise — CHANGES_REQUESTED** [review]
Maybe I am somewhat confused on what your trying to create here. Are you wanting the method to not have to require any trait implementation?

File: graph-http/src/types/nextlink.rs
Comment (sreeise) [inline graph-http/src/types/nextlink.rs:8]: This is what I was thinking as far as a trait goes using the previous `json` method changes I suggested. With this you will have to remove `next_link()` method from `ODataLink` replace all calls of `ODataLink` where` next_link()` is used with `ODataNextLink` but thats fine. 

```suggestion
pub trait ODataNextLink {
    fn next_link(&self) -> Option<String>;
}

impl ODataNextLink for serde_json::Value {
    fn next_link(&self) -> Option<String> {
        self["@odata.nextLink"].as_str().map(|s| s.to_string())
    }
}
```

**Review by sreeise — CHANGES_REQUESTED** [review]

File: graph-http/src/types/nextlink.rs
Comment (DevLazio) [reply graph-http/src/types/nextlink.rs:8]: Ok, I understand now. There was some quiproquo between us on what `V` was. 
I thought you meant the singular item (ex `User`) but you meant a struct containing a Vec<> of items (ex `Users`). That's why I was confused about the paging. 
Well, here goes my english proficiency.

The problem with this solution is that we still require the caller to declare a struct and implement the trait manually,  that's why we went the macro way. I find it cumbersome to have a bunch of `User`, `Users`, `impl ODataNextLink for Users`, etc for each struct that I fetch from graph_api when I just want to get every `User` in my tenant and do my business logic on it. 

I feel like if we want to return a `Vec<Users>`  we should declare ourselves the `Users` (with a more generic name ofc) struct with all the possible attributes and methods (ODataLink ? ) to save the caller the trouble. And then implement a method so the caller can flatten this `Vec<Users>` to `Vec<User>`.

```rust
impl<'a> IntoResponseAsync<'a, NextLink> {
    pub async fn json<V>(self) -> GraphResult<GraphResponse<Vec<NextLinkValues<V>>>>
        where
                for<'de> V: serde::Deserialize<'de>,
    {
        if self.error.is_some() {
            return Err(self.error.unwrap_or_default());
        }
        let mut values = vec![];
        let request = self.client.build().await;
        let response = request.send().await?;
        let headers = response.headers().clone();
        let status = response.status();
        let url = GraphUrl::from(response.url());
        let body: NextLinkValues<V> = response.json().await.map_err(GraphFailure::from)?;
        let mut next_link = body.next_link.clone().and_then(|url| GraphUrl::parse(&url).ok());
        values.push(body);
        while let Some(url) = next_link {
            self.client.set_url(url);
            let request = self.client.build().await;
            let response = request.send().await?;
            let body: NextLinkValues<V> = response.json().await.map_err(GraphFailure::from)?;
            next_link = body.next_link.clone().and_then(|url| GraphUrl::parse(&url).ok());
            values.push(body);
        }
        Ok(GraphResponse::new(url, values, status, headers))
    }
}

#[derive(Deserialize)]
pub struct NextLinkValues<V> {
    pub value: Vec<V>,
    #[serde(rename = "@odata.nextLink")]
    pub next_link: Option<String>,
}

pub trait Foo<V> {
    fn to_v(self) -> Vec<V>;
}

impl<V> Foo<V> for Vec<NextLinkValues<V>> {
    fn to_v(self) -> Vec<V> {
        let mut b = vec![];

        for mut c in self.into_iter().map(|a| a.value) {
            b.append(&mut c)
        }
        b
    }
}
```

so the caller can do something like that :
image

Comment (sreeise) [reply graph-http/src/types/nextlink.rs:8]: > There was some quiproquo between us on what V was.

I have no idea what `quiproquo` means 😆. If you are saying quid pro quo, which is an actual English saying, then that is not what it means 😄.  But its fine I think the word to use there should just be miscommunication.

There was some miscommunication between us on what V was.

**Quid pro quo** just means a favor or advantage granted or expected in return for something

Whereas **miscommunication** means failure to communicate adequately.

I am honestly not trying to be rude or mean by explaining those words. I genuinely hope that it helps you.

> The problem with this solution is that we still require the caller to declare a struct and implement the trait manually, that's why we went the macro way. I find it cumbersome to have a bunch of User, Users, impl ODataNextLink for Users, etc for each struct that I fetch from graph_api when I just want to get every User in my tenant and do my business logic on it.'

I can see where your coming from and I get it. But I also know that this is something where people have different opinions on what they want and every use case is different. Some people do want to declare a User and Users struct and have a way to implement traits such as this. This is why I mentioned that your solution is great for those who just want to get all of the users regardless. But that doesn't take into account other use cases. 

As a library maintainer my goal is to provide the ability to do things the way each person wants. There are libraries that take a very opinionated approach to things but I typically have tried not to.

> I feel like if we want to return a Vec<Users> we should declare ourselves the Users (with a more generic name ofc) struct with all the possible attributes and methods (ODataLink ? ) to save the caller the trouble. And then implement a method so the caller can flatten this Vec<Users> to Vec<User>.

I have no problem with us flattening the `Vec<Users>` to just a `Vec<User>`. My example wasn't meant to say that the `Vec` should be a `Vec<Users>`. I was just trying to explain with an example what I meant about the trait. 👍 


Ok with all of that said, I think your probably fine staying with what you have now. Except maybe lets clean up some of the code a bit like you have in your reply there. Meaning lets use the `while let loop`. You don't need to do a to_v method. You can just return the vec flattened completley. 

And I dont know if this was intentional, but you can just use the `flatten()` method on an iterator of `Vec<Vec<T>>` in order to turn it into a `Vec<T>`.

Comment (DevLazio) [reply graph-http/src/types/nextlink.rs:8]: > > There was some quiproquo between us on what V was.
> 
> I have no idea what `quiproquo` means 😆. If you are saying quid pro quo, which is an actual English saying, then that is not what it means 😄. But its fine I think the word to use there should just be miscommunication.
> 
> There was some miscommunication between us on what V was.
> 
> **Quid pro quo** just means a favor or advantage granted or expected in return for something
> 
> Whereas **miscommunication** means failure to communicate adequately.
> 
> I am honestly not trying to be rude or mean by explaining those words. I genuinely hope that it helps you.

Sorry, I got confused with the french word [quiproquo](https://www.linguee.com/french-english/translation/quiproquo.html). Thanks for the explanations !

> Ok with all of that said, I think your probably fine staying with what you have now. Except maybe lets clean up some of the code a bit like you have in your reply there. Meaning lets use the `while let loop`. You don't need to do a to_v method. You can just return the vec flattened completley.

Sure. I'll do that.

> And I dont know if this was intentional, but you can just use the `flatten()` method on an iterator of `Vec<Vec<T>>` in order to turn it into a `Vec<T>`. Again im not sure if you knew that already or if there was some other reason for the method you wrote in your reply.
> 
> An example of flatten:
> 
> ```rust
>     let mut vec1: Vec<Vec<&str>> = vec![
>         vec!["a", "b", "c"],
>         vec!["e", "f", "g"]
>     ];
> 
>     let vec2: Vec<&str> = vec1.into_iter().flatten().collect();
>     println!("{vec2:?}"); // -> ["a", "b", "c", "e", "f", "g"]
> ```

Huh. I did know about it. I remember trying it before the `map` but not after. 
It is indeed much cleaner: 
``` rust
impl<V> Foo<V> for Vec<NextLinkValues<V>> {
    fn to_v(self) -> Vec<V> {
        self.into_iter().map(|a| a.value).flatten().collect()
    }
}
```

**Comment by sreeise** [comment]
@DevLazio Thank you for sticking with this. Great work and I appreciate it.

**Pull request merged**
